go-controller/pkg/node/gateway_init.go

```diff
@@ -259,59 +259,86 @@ func configureSvcRouteViaInterface(iface string, gwIPs []net.IP) error {
 func (nc *DefaultNodeNetworkController) initGateway(subnets []*net.IPNet, nodeAnnotator kube.Annotator,
 	waiter *startupWaiter, managementPortConfig *managementPortConfig, kubeNodeIP net.IP) error {
 	klog.Info("Initializing Gateway Functionality")
+
+	// filter out unsupported modes
+	if config.OvnKubeNode.Mode != types.NodeModeFull && config.Gateway.Mode != config.GatewayModeShared {
+		return fmt.Errorf("combination of gateway mode %#v and node mode %#v not supported",
+			config.Gateway.Mode,
+			config.OvnKubeNode.Mode)
+	}
+
 	var err error
+	var egressGWInterface string
 	var ifAddrs []*net.IPNet
-
 	var loadBalancerHealthChecker *loadBalancerHealthChecker
 	var portClaimWatcher *portClaimWatcher
 
-	if config.Gateway.NodeportEnable && config.OvnKubeNode.Mode == types.NodeModeFull {
-		loadBalancerHealthChecker = newLoadBalancerHealthChecker(nc.name, nc.watchFactory)
-		portClaimWatcher, err = newPortClaimWatcher(nc.recorder)
+	if config.OvnKubeNode.Mode == types.NodeModeDPUHost {
+		// On DPU host, force gateway interface to be the interface associated
+		// with kubeNodeIP
+		gatewayIntf, err := getInterfaceByIP(kubeNodeIP)
 		if err != nil {
 			return err
 		}
+		config.Gateway.Interface = gatewayIntf
 	}
 
 	gatewayNextHops, gatewayIntf, err := getGatewayNextHops()
 	if err != nil {
 		return err
 	}
 
-	egressGWInterface := ""
-	if config.Gateway.EgressGWInterface != "" {
-		egressGWInterface = interfaceForEXGW(config.Gateway.EgressGWInterface)
-	}
-
 	ifAddrs, err = getNetworkInterfaceIPAddresses(gatewayIntf)
 	if err != nil {
 		return err
 	}
 
-	// For DPU need to use the host IP addr which currently is assumed to be K8s Node cluster
-	// internal IP address.
 	if config.OvnKubeNode.Mode == types.NodeModeDPU {
+		// For DPU need to use the host IP addr which currently is assumed to be
+		// K8s Node cluster internal IP address.
 		ifAddrs, err = getDPUHostPrimaryIPAddresses(kubeNodeIP, ifAddrs)
 		if err != nil {
 			return err
 		}
+	} else if config.Gateway.NodeportEnable {
+		loadBalancerHealthChecker = newLoadBalancerHealthChecker(nc.name, nc.watchFactory)
+		portClaimWatcher, err = newPortClaimWatcher(nc.recorder)
+		if err != nil {
+			return err
+		}
 	}
 
-	if err := util.SetNodePrimaryIfAddrs(nodeAnnotator, ifAddrs); err != nil {
-		klog.Errorf("Unable to set primary IP net label on node, err: %v", err)
+	if config.OvnKubeNode.Mode != types.NodeModeDPUHost {
+		if err = nc.validateVTEPInterfaceMTU(); err != nil {
+			return fmt.Errorf("failed to bvalidate VTEP interface MTU, err: %v", err)
+		}
+
+		if err := util.SetNodePrimaryIfAddrs(nodeAnnotator, ifAddrs); err != nil {
+			return fmt.Errorf("unable to set primary IP net label on node, err: %v", err)
+		}
+
+		if config.Gateway.EgressGWInterface != "" {
+			egressGWInterface = interfaceForEXGW(config.Gateway.EgressGWInterface)
+		}
```

Comment:
You can move this code inside gatewayInitInternal and drop this `if` block, which will also reduce number of arguments given to the functions